### PR TITLE
HUGE VAMPIRE UPDATE

### PR DESCRIPTION
**Pull request by VelardAmakar**
- Enthrall can be used for free, but you need at least 300 blood in your system to use it
- Full power(now "Maturity" is activated at 450 instead of 500
- Blood can be used to unholy tiles outside of the chapel.
- Coffins now heal faster and cure Nullify and Smite
- Nullrod completely ruins the vampire's day if the chaplain is using it, don't even try to take him on hand to hand
- Being burned when dead reduces your corpse to ashes
- Holywater fucks you up almost as bad as the sun, but slower
- Chapel kills a weak vampire almost as fast as the sun
- Sunlight is even more lethal
- Picking up a bible or a nullrod will smite you(but not enough to hurt you, trying to use them does that)
- NEW POWER: Charisma, activates at 450 blood, increases Hypnotise & Glare stun time, and greatly decreases Enthrall time/cooldown.
- NEW POWER: Shadowy Menace, activates at 450 blood, buffs Shadow Cloak, grants immunity to blessed tiles(but not in the chapel) and gives a chance to terrify anyone who sees you.
- NEW POWER: Undying Lord of the Night, unlocks after 666 blood, grants the ability to revive after death(but not if you've been smote or gibbed) and spawn a free vampire cloak that can hold a few items, and has moderately decent defense against bullets, fire, and holy effects.
- Various typos and poor grammar in shit fixed, also replacing "warning"(italic red text) with the "sinister"(spooky purple text) in vamp power text blurbs
- Added check_holy() to mob_helpers.dm, which checks if a mob is on a holy tile
- Moved coffin healing vampires to vampire.dm where it belongs
- Adds new smite debuff that is increased by holy things, but decreased by vampire powers and sleeping in coffins.

### CODE REVIEW

**Comment by VelardAmakar** [comment]
yes it's all one commit
yes it's poor form to do this
yes i'm aware that i could have done it in a less sloppy way
i apologize, there's just so much that relies on other files and the pace i was working at made doing separate commits kind of useless
i'll do better in the future

i'll have changelogs in as soon as a maintainer looks at this and gives it the thumbs up

File: code/game/gamemodes/vampire/vampire.dm [hunk 1]
```diff
@@ -270,14 +273,14 @@ You are weak to holy things and starlight. Don't go into space and avoid the Cha
 	src.attack_log += text("\[[time_stamp()]\] <font color='red'>Bit [H.name] ([H.ckey]) in the neck and draining their blood</font>")
 	H.attack_log += text("\[[time_stamp()]\] <font color='orange'>Has been bit in the neck by [src.name] ([src.ckey])</font>")
 	log_attack("[src.name] ([src.ckey]) bit [H.name] ([H.ckey]) in the neck")
-	src.visible_message("\red <b>[src.name] bites [H.name]'s neck!<b>", "\red <b>You bite [H.name]'s neck and begin to drain their blood.", "\blue You hear a soft puncture and a wet sucking noise")
+	src.visible_message("<span class='danger'>[src.name] bites [H.name]'s neck!</span>", "<span class='danger'>You bite [H.name]'s neck and begin to drain their blood.</span>", "<span class='notice'>You hear a soft puncture and a wet sucking noise.</span>")
```
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire.dm]: Try to split long calls like this across lines, eg:
```dm
src.visible_message(
    "<span class='danger'>[src.name] bites [H.name]'s neck!</span>", 
    "<span class='danger'>You bite [H.name]'s neck and begin to drain their blood.</span>", 
    "<span class='notice'>You hear a soft puncture and a wet sucking noise.</span>")
```

File: code/game/gamemodes/vampire/vampire.dm [hunk 2]
```diff
@@ -351,10 +354,25 @@ You are weak to holy things and starlight. Don't go into space and avoid the Cha
 		if(!(VAMP_SLAVE in vamp.powers))
 			vamp.powers.Add(VAMP_SLAVE)
 
-	// TIER 5
+	// TIER 5 (/vg/)
+	if(vamp.bloodtotal >= 400)
+		if(!(VAMP_MATURE in vamp.powers))
+			vamp.powers.Add(VAMP_MATURE)
```
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire.dm]: You can simplify this to ```vamp.powers |= VAMP_MATURE```, which will add it to powers if it's not there already

File: code/setup.dm [hunk 3]
```diff
@@ -864,19 +864,22 @@ var/list/RESTRICTED_CAMERA_NETWORKS = list( //Those networks can only be accesse
...
+#define VAMP_CLOAK    7
+#define VAMP_BATS     8
+#define VAMP_SCREAM   9
+#define VAMP_JAUNT    12
+#define VAMP_SLAVE    13
+#define VAMP_BLINK    14
+#define VAMP_MATURE   15
+#define VAMP_SHADOW   16
+#define VAMP_CHARISMA 17
+#define VAMP_UNDYING  18
```
Comment (wild-billy) [inline code/setup.dm]: Consider defining a VAMP_MAX indicating the highest value: see above for why

File: code/game/gamemodes/vampire/vampire.dm [hunk 4]
```diff
@@ -363,35 +381,46 @@ You are weak to holy things and starlight. Don't go into space and avoid the Cha
 		if(!(n in old_powers))
 			switch(n)
```
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire.dm]: This switch is pretty bad, can improve it by defining something like this:
```dm
var/list/allPowers[VAMP_MAX] //See below
allPowers[VAMP_SHAPE] =list(
    "gainText" = "You have gained the shapeshifting ability, at the cost of stored blood you can change your form permanently.",
    "verbPath" = /client/proc/vampire_shapeshift)
[...] //Define the rest
```dm
Then that huge switch just turns into:
```dm
if(!(n in old_powers))
    src << allPowers[n]["gainText"]
    verbs += allPowers[n]["verbPath"]
```
A better solution would be to make them datums (the above are faux objects with faux vars).
Then the code would look like:
```dm
/vampower
    var/gainText = "" //Or something basic
    var/verbPath = /client/proc/vampire_error //Or something basic

/vampower/proc/bestow(mob/player/target)
    if(target isn't a vampire) return 0
    if(target isn't powerful enough) return 0
    if(target already has this power) return 0
    target << src.gainText
    target.verbs |= src.verbPath

/vampower/shape
    gainText = "You have gained the shapeshifting ability, at the cost of stored blood you can change your form permanently.",
    verbPath = /client/proc/vampire_shapeshift)

[...] //Define the rest
```
But it looks like it'd take quite a bit of effort to refactor the current code to that.

File: code/game/gamemodes/vampire/vampire.dm [hunk 5]
```diff
@@ -494,6 +525,75 @@ You are weak to holy things and starlight. Don't go into space and avoid the Cha
...
 
+/mob/living/carbon/human/proc/handle_vampire_smite()
+	var/smitetemp = 0
+	var/vampcoat = istype(wear_suit, /obj/item/clothing/suit/storage/draculacoat) //coat reduces smiting
+	if(check_holy()) //if you're on a holy tile get ready for pain
+		smitetemp += (vampcoat ? 1 : 5)
+		if(prob(35))
+			src << "<span class='danger'>This ground is blessed. Get away, or splatter it with blood to make it safe for you.</span>"
+
+	if(istype(loc.loc, /area/chapel)) //stay out of the chapel unless you want to turn into a pile of ashes
```
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire.dm]: don't use ```loc.loc```, if you want an area use ```get_area()```
```loc.loc``` is not guaranteed to be an area
eg it they're in a locker, it'll be a turf

File: code/game/gamemodes/vampire/vampire.dm [hunk 6]
```diff
@@ -494,6 +525,75 @@ You are weak to holy things and starlight. Don't go into space and avoid the Cha
...
+		smitetemp += (vampcoat ? 5 : 15)
+		if(prob(35))
+			src << "<span class='sinister'>Burn, wretch.</span>"
+
+	if(!mind.vampire.nullified) //Checks to see if you can benefit from your vamp powers here
+		if(VAMP_MATURE in mind.vampire.powers)
+			smitetemp -= 1
+		if(VAMP_SHADOW in mind.vampire.powers)
+			var/turf/simulated/T = get_turf(src)
+			if(T.lighting_lumcount > 2)
```
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire.dm]: Double check this, make sure it's supposed to be ```>``` and not ```<```

File: code/game/gamemodes/vampire/vampire.dm [hunk 7]
```diff
@@ -503,6 +603,14 @@ You are weak to holy things and starlight. Don't go into space and avoid the Cha
 		hud_used.vampire_blood_display.maptext_height = 64
 		hud_used.vampire_blood_display.maptext = "<div align='left' valign='top' style='position:relative; top:0px; left:6px'> U:<font color='#33FF33' size='1'>[mind.vampire.bloodusable]</font><br> T:<font color='#FFFF00' size='1'>[mind.vampire.bloodtotal]</font></div>"
 	handle_vampire_cloak()
+	handle_vampire_menace()
+	handle_vampire_smite()
 	if(istype(loc, /turf/space))
 		check_sun()
+	if(istype(loc, /obj/structure/closet/coffin))
```
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire.dm]: This is snowflakey, consider putting a proc or var on structures indicating their interactions with vampires

File: code/game/gamemodes/vampire/vampire_powers.dm [hunk 8]
```diff
@@ -33,13 +33,19 @@
 		if(!fullpower)
 			src << "<span class='warning'>Your powers are useless on this holy ground.</span>"
 			return 0
+	if(check_holy())
+		if(!fullpower)
+			var/turf/simulated/T = get_turf(src)
+			if(T.lighting_lumcount > 2)
```
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire_powers.dm]: This threshold (```2```) appears multiple times, make it a define (or var, or config option, etc).
Rule of thumb is that any literal (bare data value) that appears more than once should be referred to with an identifier
Better for clarity, maintainability, etc

File: code/game/gamemodes/vampire/vampire_powers.dm [hunk 9]
```diff
@@ -93,6 +99,102 @@
 		spawn(200)
 			M.current.verbs += /client/proc/vampire_rejuvinate
 
+/client/proc/vampire_returntolife()
```
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire_powers.dm]: Yikes. This proc should call something in medical code to do the resurrection, half this code isn't vampire-specific at all

File: code/game/gamemodes/vampire/vampire_powers.dm [hunk 10]
```diff
@@ -194,7 +296,7 @@
 			C.Stun(distance_value)
 			if(distance_value > 1)
 				C.Weaken(distance_value)
-			C.stuttering += 5+distance_value
+			C.stuttering += 5+distance_value * (VAMP_CHARISMA in M.vampire.powers ? 2 : 1) //double stutter time with Charisma
```
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire_powers.dm]: Always stick parens around use of ```in```

File: code/game/gamemodes/vampire/vampire_powers.dm [hunk 11]
```diff
@@ -241,26 +343,25 @@
...
+	if(M.current.can_enthrall(C) && do_mob(M.current, C, VAMP_CHARISMA in M.vampire.powers ? 25 : 50)) //takes half the time with Charisma unlocked
 		if(M.current.can_enthrall(C) && M.current.vampire_power(300, 0)) // recheck
 			M.current.handle_enthrall(C)
-			M.current.remove_vampire_blood(300)
 			M.current.verbs -= /client/proc/vampire_enthrall
-			spawn(1800) M.current.verbs += /client/proc/vampire_enthrall
+			spawn(VAMP_CHARISMA in M.vampire.powers ? 600 : 1800) M.current.verbs += /client/proc/vampire_enthrall
 		else
-			M.current << "<span class='warning'>You or your target either moved or you dont have enough usable blood.</span>"
+			M.current << "<span class='warning'>Either you or your target moved, or you aren't strong enough at the moment.</span>"
```
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire_powers.dm]: Try to disentangle the above and return a more specific message

File: code/game/gamemodes/vampire/vampire_powers.dm [hunk 12]
```diff
@@ -331,7 +440,7 @@
 	ticker.mode.enthralled.Add(H.mind)
 	ticker.mode.enthralled[H.mind] = src.mind
 	H.mind.special_role = "VampThrall"
-	H << "<b><span class='warning'>You have been Enthralled by [name]. Follow their every command.</b></span>"
+	H << "<b><span class='sinister'>You have been Enthralled by [name]. Follow their every command.</b></span>"
```
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire_powers.dm]: The ```<b>``` shouldn't be there, that's what the span classes are for

File: code/game/objects/items/weapons/storage/bible.dm [hunk 13]
```diff
@@ -61,19 +61,33 @@
 	log_attack("<font color='red'>[user.name] ([user.ckey]) attacked [M.name] ([M.ckey]) with [src.name] (INTENT: [uppertext(user.a_intent)])</font>")
 
 	if (!(istype(user, /mob/living/carbon/human) || ticker) && ticker.mode.name != "monkey")
```
Comment (wild-billy) [inline code/game/objects/items/weapons/storage/bible.dm]: Gah this damn check, great example of what not to do

File: code/game/objects/items/weapons/weaponry.dm [hunk 14]
```diff
@@ -55,9 +55,11 @@
 	if(M.mind)
 		if(M.mind.vampire)
 			if(ishuman(M))
-				if(!(VAMP_FULL in M.mind.vampire.powers))
+				if(!(VAMP_MATURE in M.mind.vampire.powers))
```
Comment (wild-billy) [inline code/game/objects/items/weapons/weaponry.dm]: Consider fixing this staircase here

File: code/game/objects/items/weapons/weaponry.dm [hunk 15]
```diff
@@ -66,6 +68,13 @@
 		user << "\blue You hit the floor with the [src]."
 		call(/obj/effect/rune/proc/revealrunes)(src)
 
+/obj/item/weapon/nullrod/pickup(mob/living/user as mob)
+	if(user.mind && user.mind.assigned_role == "Chaplain")
+		user << "<span class ='notice'>The obsidian rod is teeming with divine power. You feel like you could pulverize a horde of undead with this.</span>"
+	if(user.mind && user.mind.vampire && (!VAMP_UNDYING in user.mind.vampire.powers))
```
Comment (wild-billy) [inline code/game/objects/items/weapons/weaponry.dm]: don't need to repeat the ```user.mind``` check

File: code/modules/clothing/suits/miscellaneous.dm [hunk 16]
```diff
@@ -418,4 +418,13 @@
...
+	item_state = "kefkarobe"
+
+/obj/item/clothing/suit/storage/draculacoat
+	name = "Vampire Coat"
+	desc = "What is a man? A miserable little pile of secrets."
+	icon_state = "draculacoat"
+	item_state = "draculacoat"
+	blood_overlay_type = "coat"
+	cant_hold = list(/obj/item/weapon/nullrod, /obj/item/weapon/storage/bible)
+	armor = list(melee = 30, bullet = 20, laser = 10, energy = 10, bomb = 0, bio = 0, rad = 0)
```
Comment (wild-billy) [inline code/modules/clothing/suits/miscellaneous.dm]: add EOL

File: code/modules/mob/living/carbon/human/human.dm [hunk 17]
```diff
@@ -584,6 +584,8 @@
 		return get_id_name("Unknown")
 	if( head && (head.flags_inv&HIDEFACE) )
 		return get_id_name("Unknown")		//Likewise for hats
+	if(mind && mind.vampire && (VAMP_SHADOW in mind.vampire.powers) && mind.vampire.ismenacing)
```
Comment (wild-billy) [inline code/modules/mob/living/carbon/human/human.dm]: this breaks encapsulation but you don't have much of a choice unless you want to refactor this code too :V

File: code/modules/mob/living/carbon/human/life.dm [hunk 18]
```diff
@@ -1422,9 +1416,9 @@ var/global/list/organ_damage_overlays = list(
 					see_in_dark = 8
 					see_invisible = SEE_INVISIBLE_LEVEL_ONE
 		if(mind && mind.vampire)
```
Comment (wild-billy) [inline code/modules/mob/living/carbon/human/life.dm]: This breaks encapsulation, you can likely fix it by adding a simple hook (well, sort of simple):
```dm
for(var/toCall in src.callOnLife)
    if(locate(toCall) && callOnLife[toCall])
        call(locate(toCall),callOnLife[toCall])()
    else callOnLife -= toCall
```
where ```callOnLife``` is a list that you add things to like this:
```dm
callOnLife += list("\ref[foo]" = "bar")
```
```foo``` is an object that's interested in doing shit every Life() tick.
```bar``` is the shit it wants to do (the name of the proc to call, the proc has to be one of foo's)
This allows you to do vamp things in Life() AND keep vamp code contained in its own files: just define a proc with everything you need to do (all the code below) and add it to ```callOnLife``` in the code that makes someone a vampire.

Note that this actually shows some positive features of DM: the above code is horrific or impossible in most other programming languages (google "smart pointers" and "function pointers" for example).

File: code/modules/mob/mob_helpers.dm [hunk 19]
```diff
@@ -29,6 +29,12 @@
 			return 1
 	return 0
 
+/proc/check_holy()
+	var/turf/T = get_turf(src)
+	if(!T)	return 0
+	if(!T.holy)	return 0
```
Comment (wild-billy) [inline code/modules/mob/mob_helpers.dm]: try to use spaces for non-syntactic alignment, many things (like github) have broken tab settings and won't display it properly with tabs

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 20]
```diff
@@ -199,6 +199,9 @@
 	else if(istype(self.data["donor"], /mob/living/carbon/alien))
 		var/obj/effect/decal/cleanable/blood/B = blood_splatter(T,self,1)
 		if(B) B.blood_DNA["UNKNOWN DNA STRUCTURE"] = "X*"
+
+	if(volume >= 5 && !istype(T.loc, /area/chapel)) //blood desanctifies non-chapel tiles
```
Comment (wild-billy) [inline code/modules/reagents/Chemistry-Reagents.dm]: Can replace this with generic holiness (see hastebin)

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 21]
```diff
@@ -685,18 +688,22 @@
 				O.show_message(text("<span class='notice'>[]'s eyes blink and become clearer.</span>", M), 1) // So observers know it worked.
 		// Vamps react to this like acid
```
Comment (wild-billy) [inline code/modules/reagents/Chemistry-Reagents.dm]: Breaks encapsulation, but it's hopelessly broken for our med/chem systems anyway.

File: code/game/gamemodes/vampire/vampire.dm (hunk 4 above)
Comment (ComicIronic) [inline code/game/gamemodes/vampire/vampire.dm]: Ideally these could use the current spell framework, and use an altered charge system to use blood.

File: code/game/objects/items/weapons/weaponry.dm (hunk 15 above)
Comment (ComicIronic) [inline code/game/objects/items/weapons/weaponry.dm]: Shouldn't the ! be around the in, not next to VAMP_UNDYING?
That is, !(VAMP_UNDYING in user.mind.vampire.powers)

File: code/modules/mob/mob_helpers.dm [hunk 22]
```diff
@@ -29,6 +29,12 @@
 			return 1
 	return 0
 
+/proc/check_holy()
```
Comment (ComicIronic) [inline code/modules/mob/mob_helpers.dm]: This should be an atom proc or take an atom argument, not be an unattached proc.
I'm honestly surprised this was allowed to compile with the use of src.

**Comment by Duny-** [comment]
I already told you I was skeptical about the whole free enthrall thing but the rest seems good 
bluethumbsup

By the way transformation to bat when, I think I saw a gif in the thread for that at some point. It would truly make vampire the changeling replacement. :^)

File: code/game/objects/items/weapons/weaponry.dm (hunk 15 above)
Comment (d3athrow) [inline code/game/objects/items/weapons/weaponry.dm]: >Shouldn't the ! be around the in, not next to VAMP_UNDYING?
That is, !(VAMP_UNDYING in user.mind.vampire.powers)

This

File: code/game/gamemodes/vampire/vampire_powers.dm (hunk 8 above)
Comment (VelardAmakar) [inline code/game/gamemodes/vampire/vampire_powers.dm]: wouldn't that require me to dig deep into the hell that is lighting code

**Comment by VelardAmakar** [comment]
going to fix more tomorrow
i am trash
aaaaaaa

**Comment by SkowronX** [comment]
Fix now.

**Comment by VelardAmakar** [comment]
haven't had time to code
will see what i can do tonight

File: code/game/gamemodes/vampire/vampire.dm [hunk 23]
```diff
@@ -318,43 +323,58 @@ You are weak to holy things and starlight. Don't go into space and avoid the Cha
...
 			vamp.powers.Add(VAMP_SHAPE)
 
 	// TIER 2
 	if(vamp.bloodtotal >= 150)
-		if(!(VAMP_CLOAK in vamp.powers))
+		if(!(vamp.powers |= VAMP_CLOAK))
 			vamp.powers.Add(VAMP_CLOAK)
-		if(!(VAMP_DISEASE in vamp.powers))
+		if(!(vamp.powers |= VAMP_DISEASE))
 			vamp.powers.Add(VAMP_DISEASE)
```
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire.dm]: the check isn't necessary if you're using ```|=```, the ```|=``` itself means "add it to the list if it isn't already there".
So this:
```
if(!(vamp.powers |= VAMP_DISEASE))
 			vamp.powers.Add(VAMP_DISEASE)
```
Should look like this:
```
vamp.powers |= VAMP_DISEASE
```

Comment (VelardAmakar) [inline code/game/gamemodes/vampire/vampire.dm]: aaaaaaaa

**Comment by VelardAmakar** [comment]
will implement callonlife tomorrow then add changelog
not gonna refactor powers or holiness or add the vampire bat
that's for another time
aaaa

**Comment by VelardAmakar** [comment]
>yes it's all one commit
AHAHAHA
HAHAHAHA

**Comment by VelardAmakar** [comment]
i'm going to wait for travis to approve these commits before i go to sleep
i swear i'm going to start tripping out if i stay up too much longer

File: code/game/gamemodes/vampire/vampire_powers.dm (hunk 8 above)
Comment (wild-billy) [inline code/game/gamemodes/vampire/vampire_powers.dm]: Nah, just gotta do something like
```
#define VAMP_LIGHT_THRESHOLD 2
```
at the top of the file, and #undef it at the bottom

**Comment by VelardAmakar** [comment]
come on travis finish checking my bullshit so it can be merged

**Comment by IratePirate** [comment]
inb4 more conflicts somehow

**Comment by VelardAmakar** [comment]
travis approves, merge when

**Comment by d3athrow** [comment]
Skimmed over it and other than the issues wb had with it, it looks good to me.

**Comment by d3athrow** [comment]
Thanks github for not detecting the manual merge, e0bc7fc1cedf18a2df9447afc1f377585e5c5861

**Pull request closed without merging**